src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.Unix.cs

```diff
@@ -0,0 +1,115 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System.Buffers;
+using System.IO;
+using System.Reflection;
+
+namespace System.Runtime.Loader
+{
+    public partial class AssemblyLoadContext
+    {
+        // Find satellite path using case insensitive culture name
+        internal static unsafe string? FindCaseInsensitiveSatellitePath(string parentDirectory, string cultureName, string assembly)
```

Comment:
I am not sure whether we need to be do the case-insensitive probing here. I know we have it in some places, but it is expensive, not done consistently and a pit of failure.

Can we just document that the case has to match?

Reply:
Actually have strict case naming causing some problems. Sometime the globalization name can be returned in different casing. so it will be hard for people to always have exact casing. one example is zh-Hans/Hant which can be also valid as zh-HANS and zh-HANT. of course we can strict casing in the globalization layer but we'll have to add a lot of special cases as we encounter them. looks we only probe for other casing only if we couldn't load the satellite assembly which I think can be acceptable perf wise.

Reply:
> Sometime the globalization name can be returned in different casing. 

Is there a default/standard casing for the culture name IDs? Can we normalize to that in the tooling? 

> Looks we only probe for other casing only if we couldn't load the satellite assembly which I think can be acceptable perf wise.

It is not just about perf. It is also about a number of corner cases that we need to deal with and that will not do what people want half of the time.

Reply:
> Is there a default/standard casing for the culture name IDs

Yes but normalizing the name would tricky. We have to know if this culture is neutral culture or not to do the right casing ourselves. en-US is the casing for the specific cultures. while zh-Hant is for neutrals.  we can know if the culture is neutral or not in main stream cases but not when we are in Invariant mode. As I mentioned we can hard code the cases we know when we are invariant mode but we expect to add more as we encounter more. 

> Can we normalize to that in the tooling?
either the tooling or at runtime, we'll have to normalize it ourselves in some cases as I mentioned in the globalization mode.

> It is also about a number of corner cases that we need to deal with and that will not do what people want half of the time.

Is there examples of that? Just for my education.

Reply:
> not when we are in Invariant mode

If you are in Invariant mode, you are saying that you do not care about globalization. This does not have to work well or at all when we are in invariant mode.

> Is there examples of that? 

For example, you can have conflicts that differ just by casing. How do you pick the winner?

Reply:
> If you are in Invariant mode, you are saying that you do not care about globalization. This does not have to work well or at all when we are in invariant mode.

Unfortunately as we ship Alpine images which have invariant mode on by default, and these images got used in some scenarios while the users not aware about that and then they run into some problems like this resource loading. 

> For example, you can have conflicts that differ just by casing. How do you pick the winner?

We probe first for the exact casing. If we couldn't find the exact casing, then the first enumerated match would be reasonable anyway because I am not expecting people will create multiple resource assemblies with exact same name. 

To summarize, I think it is nice to support case insensitive culture names as a resource directories. If we have any complication with that, then I am ok not support it and clearly document the behavior.

Reply:
To try to minimize corner cases, when searching for case insensitive matches, I don't stop on the first match, but parse the whole directory. If there is more than one matching directory, I ignore them all.

Reply:
This enumeration still does not feel right. We own the app publish format and can publish stuff in whatever format works best for runtime, e.g. we can lower-case the culture names and be done with it.

Reply:
> lower-case the culture names 

It would certainly make the culture lookup easier. It definitely is painful to do this sort of lookup.

Backwards compat?
.NET Standard compat?

Mayber @tarekgh's **"ok not support it and clearly document the behavior."** is better.

Reply:
The satellite probing for default context in [BindSatelliteResourceByResourceRoots](https://github.com/dotnet/coreclr/blob/3b47496ef2656333684526cf71478253a207c1a5/src/binder/assemblybinder.cpp#L1248) does not seem to doing any case-insensitive probing either. Is there any case-insensitive probing for default context hidden somewhere else?

Reply:
I don't think so.  This was added in response to @tarekgh's feature request to support case insensitive lookup of culture names on all platforms. 

I still haven't tested, but I assume the Default context hits this code too to fire it `Resolving` event. Otherwise, I'll have to add that code too...

Reply:
> We own the app publish format and can publish stuff in whatever format works best for runtime, e.g. we can lower-case the culture names and be done with it.

If we did that, I guess we could support case exact match and lowercase match for back compat….

I am drafting the patch to remove case insensitive search.

Reply:
Remove case insensitive search and opened #24211